### ISSUE TITLE
netdata  redis send wrong alarm on rdb_last_bgsave_status

### ISSUE THREAD

**Issue by yasharne** [netdata/netdata#8041]
##### Bug report summary
netdata on redis instances fires wrong alarm on rdb_last_bgsave_status but the status on the charts is ok, both alarm and tha chart is visible in the screenshot below: 
![image](https://user-images.githubusercontent.com/7656049/74244493-c476b080-4cf6-11ea-8399-96e4fb6afbbb.png)
here is the output of `info` on the `redis-cli` command:
![image](https://user-images.githubusercontent.com/7656049/74244568-e112e880-4cf6-11ea-94ee-14170c85db8f.png)

##### OS / Environment

```
/etc/lsb-release:DISTRIB_ID=Ubuntu
/etc/lsb-release:DISTRIB_RELEASE=16.04
/etc/lsb-release:DISTRIB_CODENAME=xenial
/etc/lsb-release:DISTRIB_DESCRIPTION="Ubuntu 16.04.6 LTS"
/etc/os-release:NAME="Ubuntu"
/etc/os-release:VERSION="16.04.6 LTS (Xenial Xerus)"
/etc/os-release:ID=ubuntu
/etc/os-release:ID_LIKE=debian
/etc/os-release:PRETTY_NAME="Ubuntu 16.04.6 LTS"
/etc/os-release:VERSION_ID="16.04"
/etc/os-release:HOME_URL="http://www.ubuntu.com/"
/etc/os-release:SUPPORT_URL="http://help.ubuntu.com/"
/etc/os-release:BUG_REPORT_URL="http://bugs.launchpad.net/ubuntu/"
/etc/os-release:VERSION_CODENAME=xenial
/etc/os-release:UBUNTU_CODENAME=xenial

```

##### Netdata version

 
netdata v1.19.0
##### Component Name

python.d

**Comment by ilyam8** [comment]
Hi @yasharne 

I tested it with the latest

```console
test@DEB10-1:~$ $(ps aux | grep -E -o "[a-zA-Z/]+netdata ") -V
netdata v1.19.0-484-g0a644338
```

I see no such problem

<img width="932" alt="Screenshot 2020-02-11 at 17 45 00" src="https://user-images.githubusercontent.com/22274335/74246866-3bf81000-4cf6-11ea-8c12-9a9ab56262c7.png">

**Comment by ilyam8** [comment]
@yasharne please update to the latest and check if the bug is there. It is worksforme 🤷‍♂

**Comment by yasharne** [comment]
I still get this false alert on the netdata version 1.21.1:
![image](https://user-images.githubusercontent.com/7656049/80362203-7bff7580-8897-11ea-9d67-112d299d1822.png)
And this is the alert rule:
```
template: redis_bgsave_broken
families: *
      on: redis.bgsave_health
   every: 10s
    crit: $rdb_last_bgsave_status != 0
   units: ok/failed
    info: states if redis bgsave is working
   delay: down 5m multiplier 1.5 max 1h
      to: dba
```
Can this be related to the #8751 ?

**Comment by ilyam8** [comment]
Probably, try to change

`crit: $rdb_last_bgsave_status != 0`

to 

`crit: $rdb_last_bgsave_status != nan AND $rdb_last_bgsave_status != 0

**Comment by yasharne** [comment]
@ilyam8 I'm not sure how this can solve the problem since the chart has value on the alert time:
![image](https://user-images.githubusercontent.com/7656049/80449807-532dbd80-8935-11ea-8676-2e5de137249a.png)
![image](https://user-images.githubusercontent.com/7656049/80449844-6771ba80-8935-11ea-815e-0ed057ae2f42.png)

**Comment by ilyam8** [comment]
Ok, it is 0 on the chart, the alarm condition is `$value != 0`, but the alarm is up. I dont know what is the problem.

**Comment by ilyam8** [comment]
@thiagoftsm ^^

**Comment by thiagoftsm** [comment]
I test this alarm on 3 different computers with different configuration using the latest Netdata and I could not have the alarm raised. All environments were Slackware current.

I am getting now the Ubuntu 16.04 to test this alarm on it.

**Comment by ilyam8** [comment]
I dont think there is much sense to test in such way, do not spend much time on it.

> Ok, it is 0 on the chart, the alarm condition is $value != 0, but the alarm is up. I dont know what is the problem.

You know the code, how can this be possible?

**Comment by thiagoftsm** [comment]
Hi @ilyam8 ,

The unique possible explanation for this happens on old system and does not happen today with the newest systems is a better comparison with double values. In the past I had problems like this where the result was Zero for our eyes, but in reality the computer was storing 1e-100.  This is a common problem that can happen when we compare double values and there are various papers talking about this.

A possible solution for this is instead we verify if the values are equal , we need to verify if the result of the subtraction is bigger than a `DBL_EPSILON`,

**Comment by ilyam8** [comment]
> Ok, it is 0 on the chart, the alarm condition is $value != 0, but the alarm is up. I dont know what is the problem.

@thiagoftsm how to check this specific problem?

Get values from `chart` and `alarm` endpoints?

**Comment by thiagoftsm** [comment]
@ilyam8 I am suspecting that it is possible we are storing a value like 0.00001, so this won't be visible on charts unless we use zoom, to confirm this, @yasharne when you have a "wrong" alarm again, please, could you get the chart data using `https://localhost:19999/api/v1/data?chart=redis_local.bgsave_health&after=-30&points=5` and attach the result in a comment? Case you know the exact timestamp, you can use overwrite `after`.

A last information today using Ubuntu 16.04 I could not get the alarms.

**Comment by yasharne** [comment]
@thiagoftsm The data seems fine:
```
{
 "labels": ["time", "rdb save"],
    "data":
 [
      [ 1588393050, 0],
      [ 1588393045, 0],
      [ 1588393040, 0],
      [ 1588393035, 0],
      [ 1588393030, 0],
      [ 1588393025, 0]
  ]
}
```
I'm not very sure but I think there is something wrong with the healthdb or dbengine on the master server (maybe #8751)

**Comment by thiagoftsm** [comment]
Was the values requested on Master? Case it was, the dbengine is storing correct values. I also cannot see any place in the health code to create a wrong alarm, because it allocates memory and reset it after the first action. 
How did you install your Netdata on this host?

**Comment by yasharne** [comment]
@thiagoftsm Yes, I'm reading the value from the master because web interface is not active on the slaves.
We have installed the netdata using the prebuilt binary installer

**Comment by thiagoftsm** [comment]
Thank you for your answer!
I will talk with our SRE team to understand better how we are building the binaries, because when I compiled directly on the VM, I could not recreate the errors.

**Comment by yasharne** [comment]
@thiagoftsm After deleting contents of `/opt/netdata/var/lib/netdata/` and `/opt/netdata/var/cache/netdata/` on the master node, the fake alarms disappeared too.

**Comment by thiagoftsm** [comment]
Thank you for the additional information @yasharne .
Probably there was something wrong with the data stored (@mfundul did you already see problems like this with the static-binaries?)

**Comment by mfundul** [comment]
> 
> 
> Thank you for the additional information @yasharne .
> Probably there was something wrong with the data stored (@mfundul did you already see problems like this with the static-binaries?)

Haven't seen this before.

**Comment by ilyam8** [comment]
> I am suspecting that it is possible we are storing a value like 0.00001

Impossible.

https://github.com/netdata/netdata/blob/5a718bb8b02eef930b41ef1f79873d93c2d38248/collectors/python.d.plugin/redis/redis.chart.py#L228

It is either `0` or `1`.

---

> After deleting contents of /opt/netdata/var/lib/netdata/ and /opt/netdata/var/cache/netdata/ on the master node, the fake alarms disappeared too.

We have fixed a bug with archived charts/health and some other related related bugs.

I am closing the issue. @mfundul @thiagoftsm if you think there is something we need to investigate feel free to reopen.

**Comment by feliperaul** [comment]
I'm also facing this on netdata v1.35.0-123-nightly

```
redis-cli info

# Persistence
loading:0
rdb_changes_since_last_save:86
rdb_bgsave_in_progress:0
rdb_last_save_time:1656673599
rdb_last_bgsave_status:ok
rdb_last_bgsave_time_sec:0
rdb_current_bgsave_time_sec:-1
rdb_last_cow_size:630784
aof_enabled:0
aof_rewrite_in_progress:0
aof_rewrite_scheduled:0
aof_last_rewrite_time_sec:-1
aof_current_rewrite_time_sec:-1
aof_last_bgrewrite_status:ok
aof_last_write_status:ok
aof_last_cow_size:0
```

Is the recommended fix `rm -rf /opt/netdata/var/lib/netdata/` and `rm -rf /opt/netdata/var/cache/netdata/` ? are these safe to do without stopping netdata first?

**Comment by entrptaher** [comment]
Faced same false alert with netdata v1.35.0-127-nightly

![image](https://user-images.githubusercontent.com/8284972/177024779-e68eae82-5ad8-4c54-9d3f-5c44bcd980c2.png)

**Comment by entrptaher** [comment]
This is continuously getting same critical error for last few days. The redis logs says it's perfectly fine.

![image](https://user-images.githubusercontent.com/8284972/177079212-8d15ea4c-c0c1-4fd8-a68e-b11801da32fd.png)

**Comment by MrZammler** [comment]
Hi all.

Could you please enable debug mode for health (in netdata.conf use `0x0000000000800000` in `debug flags`), restart netdata, wait for the alert to be raised, and then share the `debug.log` to `manolis@netdata.cloud` ?

If also possible, can you check the local agent dashboard when the alarm is raised to see if the alert is also raised there?

Thanks!

**Comment by ilyam8** [comment]
@entrptaher, thanks for reporting the problem. Fixed in #13296.
